core/modules/link/link.module

```diff
@@ -75,6 +77,18 @@ function link_field_instance_settings_form($field, $instance) {
     '#description' => t('If checked, the URL field is optional and submitting a title alone will be acceptable. If the URL is omitted, the title will be displayed as plain text.'),
   );
 
+  $form['type'] = array(
```

Comment:
This form item should probably be (somehow) connected with `$form['validate_url']` above. Maybe use "#states"?

Reply:
As above.